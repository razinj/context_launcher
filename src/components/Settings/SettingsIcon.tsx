// React
import React, { useContext } from 'react'
// React Native
import { Pressable, PressableAndroidRippleConfig, StyleSheet, View } from 'react-native'
// Contexts
import GlobalContext from '../../contexts/GlobalContext'
// Icons
import Icon from 'react-native-vector-icons/MaterialCommunityIcons'
// Constants
import { PRIMARY_HEX_COLOR, SECONDARY_HEX_COLOR } from '../../constants'

const rippleConfig: PressableAndroidRippleConfig = {
  color: SECONDARY_HEX_COLOR,
  borderless: true,
  foreground: true,
  radius: 20,
}

const SettingsIcon = () => {
  const { displaySettingsBottomSheet } = useContext(GlobalContext)

  return (
    <View style={styles.wrapper}>
      <Pressable
        testID='settings-button'
        onPress={displaySettingsBottomSheet}
        android_disableSound={true}
        android_ripple={rippleConfig}>
        <Icon name='dots-vertical' size={34} style={styles.icon} color={PRIMARY_HEX_COLOR} />
      </Pressable>
    </View>
  )
}

const styles = StyleSheet.create({
  wrapper: {
    alignItems: 'center',
    flexDirection: 'row',
  },
  icon: {
    margin: 5,
  },
})

export default SettingsIcon
