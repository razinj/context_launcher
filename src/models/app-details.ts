export type AppDetails = {
  label: string
  name: string
}

export type AppDetailsWithOptionalIcon = AppDetails & { icon?: string }

export type AppDetailsWithIcon = AppDetails & { icon: string }
