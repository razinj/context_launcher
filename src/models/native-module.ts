export interface AppsModuleInterface {
  getConstants(): AppsModuleConstants
  launchApplication(packageName: string): void
  showApplicationDetails(packageName: string): void
  requestApplicationUninstall(packageName: string): void
  getApplications(callback: (applications: string) => void): void
  getApplicationIcon(packageName: string, callback: (nativeAppIcon: string) => void): void
}

type AppsModuleConstants = {
  appVersion: string
  buildNumber: string
  packageName: string
}
