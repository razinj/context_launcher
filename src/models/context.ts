// React
import { RefObject } from 'react'
// React Native
import { TextInput } from 'react-native'
// BottomSheet
import { BottomSheetModal } from '@gorhom/bottom-sheet'
// Models
import { AppDetailsWithOptionalIcon } from '../models/app-details'

export type GlobalContextType = {
  // All apps
  hideAllApps: () => void
  displayAllApps: boolean
  toggleDisplayAllApps: () => void
  // Settings bottom sheet
  settingsBottomSheetRef: RefObject<BottomSheetModal> | null
  displaySettingsBottomSheet: () => void
  // App item menu
  appItemMenuDetails: AppDetailsWithOptionalIcon | null
  setAppItemMenuDetails: (appDetails: AppDetailsWithOptionalIcon) => void
  appItemMenuBottomSheetRef: RefObject<BottomSheetModal> | null
  displayAppItemMenuBottomSheet: () => void
  // Favorite apps sort
  sortableFavoriteApps: boolean
  toggleSortableFavoriteApps: () => void
  // Global app launch procedure
  globalAppLaunchProcedure: () => void
  // Misc
  dismissKeyboard: () => void
}

export type SearchContextType = {
  // Search input
  searchInputRef: RefObject<TextInput> | null
  searchAppLaunchProcedure: () => void
  // Search query validity
  invalidCharacters: boolean
  setInvalidCharacters: (isInvalid: boolean) => void
}
