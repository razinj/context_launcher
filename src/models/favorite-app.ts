import { AppDetails } from './app-details'

export type FavoriteApp = AppDetails
