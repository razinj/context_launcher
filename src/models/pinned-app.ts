// Models
import { AppDetailsWithIcon } from './app-details'

export type PinnedApp = AppDetailsWithIcon & {
  isPermanent: boolean
  isTemporary: boolean
}

export type TemporaryPinnedAppsConfig = {
  startDate?: string
  endDate?: string
}

export type PinnedAppsState = {
  list: PinnedApp[]
  temporaryAppsConfig: TemporaryPinnedAppsConfig
}
