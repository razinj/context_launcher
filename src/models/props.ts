// React
import { ReactNode } from 'react'
// React Native
import { StyleProp, TextStyle, ViewStyle } from 'react-native'
import { BaseAnimationBuilder } from 'react-native-reanimated'
// Models
import { AppDetails } from './app-details'
import { RenderedIn } from './rendered-in'

export type AppItemProps = {
  appDetails: AppDetails
  renderedIn: RenderedIn
  appIcon?: string
  wrapperStyle?: StyleProp<ViewStyle>
  pressableStyle?: StyleProp<ViewStyle>
}

export type HighlightTextProps = {
  text: string
}

export type SearchContextWrapperProps = {
  children: ReactNode
}

export type GlobalContextWrapperProps = {
  children: ReactNode
}

export type AllAppsLetterIndexProps = {
  onPress: (letterIndex: number) => void
}

export type SettingsItemLabelProps = {
  title: string
  description?: string
  titleStyle?: StyleProp<TextStyle>
  wrapperStyle?: StyleProp<ViewStyle>
}

export type ToggleSettingsProps = SettingsItemLabelProps & {
  children: ReactNode
}

export type CustomViewProps = {
  children: ReactNode
  style?: StyleProp<ViewStyle>
  exitAnimation?: BaseAnimationBuilder | typeof BaseAnimationBuilder
  entryAnimation?: BaseAnimationBuilder | typeof BaseAnimationBuilder
}

export type CustomIconProps = {
  name: string
  size: number
  color: string
  style?: Record<string, string | number>
}
