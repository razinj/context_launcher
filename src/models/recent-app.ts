import { AppDetails } from './app-details'

export type RecentApp = AppDetails // TODO: Remove this type?
