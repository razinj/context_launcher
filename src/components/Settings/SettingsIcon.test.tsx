import { fireEvent, screen } from '@testing-library/react-native'
import { defaultGlobalContextValue } from '../../../utils/test/data'
import { renderWithProvider, renderWithProviderAndContexts } from '../../../utils/test/utils'
import { GlobalContextType } from '../../models/context'
import SettingsIcon from './SettingsIcon'

describe('<SettingsIcon /> Tests', () => {
  it('should render correctly and match snapshot', () => {
    renderWithProvider(<SettingsIcon />)

    expect(screen.toJSON()).toMatchSnapshot()
    expect(screen.getByTestId('settings-button')).toBeDefined()
  })

  it('should call function to display settings bottom sheet when pressed', () => {
    const displaySettingsBottomSheetFn = jest.fn()

    const customGlobalContextValue: GlobalContextType = {
      ...defaultGlobalContextValue,
      displaySettingsBottomSheet: displaySettingsBottomSheetFn,
    }

    renderWithProviderAndContexts(<SettingsIcon />, {
      globalContextValue: customGlobalContextValue,
    })

    const settingsButton = screen.getByTestId('settings-button')

    expect(settingsButton).toBeDefined()

    fireEvent.press(settingsButton)

    expect(displaySettingsBottomSheetFn).toBeCalled()
  })
})
