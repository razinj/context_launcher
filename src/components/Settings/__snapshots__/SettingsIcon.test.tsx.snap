// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsIcon /> Tests should render correctly and match snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "transparent",
          "borderRadius": 0,
          "height": 46,
          "margin": 0,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
          "width": 46,
        }
      }
      testID="settings-button-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderColor": "rgba(121, 116, 126, 1)",
            "borderRadius": 0,
            "borderWidth": 0,
            "elevation": 0,
            "flex": 1,
            "overflow": "hidden",
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="settings-button-container"
      >
        <View
          accessibilityComponentType="button"
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityTraits="button"
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          centered={true}
          collapsable={false}
          focusable={true}
          hitSlop={
            {
              "bottom": 6,
              "left": 6,
              "right": 6,
              "top": 6,
            }
          }
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              [
                {
                  "alignItems": "center",
                  "flexGrow": 1,
                  "justifyContent": "center",
                },
                {
                  "borderRadius": 0,
                },
              ],
            ]
          }
          testID="settings-button"
        >
          <Text
            accessibilityElementsHidden={true}
            allowFontScaling={false}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              [
                {
                  "color": "#42855B",
                  "fontSize": 30,
                },
                [
                  {
                    "lineHeight": 30,
                    "transform": [
                      {
                        "scaleX": 1,
                      },
                    ],
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                ],
                {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            ó°‡™
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
